sql/core/v1.2.1/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

```diff
@@ -143,145 +117,244 @@ private[sql] object OrcFilters extends OrcFiltersBase {
+   * the remaining nodes.
+   *
+   * The alternative approach of checking-while-building can (and did) result
+   * in exponential complexity in the height of the tree, causing perf problems with Filters with
+   * as few as ~35 nodes if they were skewed.
    */
-  private def buildSearchArgument(
-      dataTypeMap: Map[String, DataType],
+  private[sql] def buildSearchArgument(
       expression: Filter,
       builder: Builder): Option[Builder] = {
-    createBuilder(dataTypeMap, expression, builder, canPartialPushDownConjuncts = true)
+    val filteredExpression: Option[Filter] = performFilter(
+      expression,
+      canPartialPushDownConjuncts = true)
+    filteredExpression.foreach(updateBuilder(_, builder))
+    filteredExpression.map(_ => builder)
   }
 
-  /**
-   * @param dataTypeMap a map from the attribute name to its data type.
-   * @param expression the input filter predicates.
-   * @param builder the input SearchArgument.Builder.
-   * @param canPartialPushDownConjuncts whether a subset of conjuncts of predicates can be pushed
-   *                                    down safely. Pushing ONLY one side of AND down is safe to
-   *                                    do at the top level or none of its ancestors is NOT and OR.
-   * @return the builder so far.
-   */
-  private def createBuilder(
-      dataTypeMap: Map[String, DataType],
-      expression: Filter,
-      builder: Builder,
-      canPartialPushDownConjuncts: Boolean): Option[Builder] = {
+  sealed trait ActionType
+  case class TrimUnconvertibleFilters(canPartialPushDownConjuncts: Boolean) extends ActionType
+  case class BuildSearchArgument(builder: Builder) extends ActionType
+
+  // The performAction method can run both the filtering and building operations for a given
+  // node - we signify which one we want with the `actionType` parameter.
+  //
+  // There are a couple of benefits to coupling the two operations like this:
+  // 1. All the logic for a given predicate is grouped logically in the same place. You don't
+  //   have to scroll across the whole file to see what the filter action for an And is while
+  //   you're looking at the build action.
+  // 2. It's much easier to keep the implementations of the two operations up-to-date with
+  //   each other. If the `filter` and `build` operations are implemented as separate case-matches
+  //   in different methods, it's very easy to change one without appropriately updating the
+  //   other. For example, if we add a new supported node type to `filter`, it would be very
+  //   easy to forget to update `build` to support it too, thus leading to conversion errors.
+  //
+  // Doing things this way does have some annoying side effects:
+  // - We need to return an `Either`, with one action type always returning a Left and the other
+  //   always returning a Right.
+  // - We always need to pass the canPartialPushDownConjuncts parameter even though the build
```

Comment:
this one is gone.

Reply:
👍 
And now, with your other suggestion, so is the other one!